python/paddle/fluid/dygraph/dygraph_to_static/return_transformer.py

```diff
@@ -0,0 +1,247 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import gast
+
+from paddle.fluid import unique_name
+from paddle.fluid.dygraph.dygraph_to_static.utils import index_in_list
+from paddle.fluid.dygraph.dygraph_to_static.break_continue_transformer import ForToWhileTransformer
+from paddle.fluid.dygraph.dygraph_to_static.variable_trans_func import create_fill_constant_node
+
+__all__ = ['ReturnTransformer']
+
+# Constant for the name of the variable which stores the boolean state that we
+# should return
+RETURN_PREFIX = '__return'
+
+# Constant for the name of the variable which stores the final return value
+RETURN_VALUE_PREFIX = '__return_value'
+
+
+class ReturnPreAnalysisVisitor(gast.NodeVisitor):
+    """
+    Visits gast Tree and pre-analyze the information about 'return'.
+    """
+
+    def __init__(self, root_node):
+        self.root = root_node
+
+        # A list to store where the current function is.
+        self.function_def = []
+
+        # Mapping from gast.FunctionDef node to the number of return statements
+        # Python allows define function inside function so we have to handle it
+        self.count_return = {}
+        self.visit(self.root)
+
+    def visit_FunctionDef(self, node):
+        self.function_def.append(node)
+        self.count_return[node] = 0
+        self.generic_visit(node)
+        self.function_def.pop()
+        return node
+
+    def visit_Return(self, node):
+        assert len(
+            self.function_def) > 0, "Found 'return' statement out of function."
+        cur_func = self.function_def[-1]
+        if cur_func in self.count_return:
+            self.count_return[cur_func] += 1
+        else:
+            self.count_return[cur_func] = 1
+        self.generic_visit(node)
+
+    def get_func_return_count(self, func_node):
+        return self.count_return[func_node]
+
+    def set_func_return_count(self, func_node, count):
+        self.count_return[func_node] = count
```

Comment:
This function `set_func_return_count ` doesn't seem to be called.

Reply:
Yes, I found it. I will remove in my next PR which supports various-length return. Indeed it is already deleted in my working station.